File: apps/ios/Sources/Litter/Bridge/GhosttyRendererBackendBridge.swift
import Foundation

/// Thin Swift implementation of the Rust-defined `TerminalRendererBackend`
/// callback interface. Holds a weak reference to the platform-side
/// `LitterGhosttyTerminal` and hops every Ghostty C call onto the main
/// thread (Ghostty's surface APIs are not thread-safe). The Rust tick task
/// invokes these methods on the shared tokio runtime.
final class GhosttyRendererBackendBridge: TerminalRendererBackend, @unchecked Sendable {
    private weak var terminal: LitterGhosttyTerminal?

    init(terminal: LitterGhosttyTerminal) {
        self.terminal = terminal
    }

    func setFocus(focused: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.setFocused(focused)
        }
    }

    func setOcclusion(occluded: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.setOcclusion(occluded)
        }
    }

    func requestRedraw() {
        // `LitterGhosttyTerminal.requestRedraw` already hops to main.
        terminal?.requestRedraw()
    }

    func applyConfigFile(path: String) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            try? terminal?.applyConfig(atPath: path)
        }
    }

    func dispatchKey(event: TerminalKeyEvent) {
        let terminal = self.terminal
        let action = Int32(GhosttyKeyTranslator.action(for: event.action))
        let litterKey = GhosttyKeyTranslator.litterKey(for: event.code)
        let mods = Int32(GhosttyKeyTranslator.mods(for: event.mods))
        let text = event.text.isEmpty ? nil : event.text
        DispatchQueue.main.async {
            _ = terminal?.dispatchKeyAction(
                action,
                key: litterKey,
                mods: mods,
                text: text,
                composing: false
            )
        }
    }

    func dispatchText(text: String, composing: Bool) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            if composing {
                terminal?.setPreeditText(text.isEmpty ? nil : text)
            } else {
                terminal?.sendText(text)
            }
        }
    }

    func dispatchPaste(bytes: Data) {
        let terminal = self.terminal
        DispatchQueue.main.async {
            terminal?.writeOutput(bytes)
        }
    }

    // Selection support — the iOS-side painted overlay + range plumbing
    // lands in a follow-up. For now we expose the read paths and return a
    // best-effort metrics estimate so word/line range math has something
    // to work with; platform UI can wire in a real overlay later.
    func readSelection() -> String? {
        // `LitterGhosttyTerminal.visibleText` doesn't differentiate the
        // active selection from the viewport. Return `nil` until the
        // overlay PR wires `ghostty_surface_read_selection`.
        nil
    }

    func readText(startRow: UInt32, startCol: UInt32, endRow: UInt32, endCol: UInt32) -> String? {
        nil
    }

    func cellMetrics() -> TerminalCellMetrics {
        TerminalCellMetrics(
            cellWidthPx: 0,
            cellHeightPx: 0,
            cols: 0,
            rows: 0,
            viewportTop: 0
        )
    }

    func setSelectionOverlay(range: TerminalCellRange?) {
        // Painted overlay arrives in a follow-up; storing the range here
        // would just be dead state. No-op until that lands.
    }
}

/// Translation: Rust `TerminalKey*` → bridge-level `LitterGhosttyKey`.
/// Bridge does the final Ghostty-enum mapping in Obj-C.
enum GhosttyKeyTranslator {
    static func action(for value: TerminalKeyAction) -> Int {
        switch value {
        case .release: return 0
        case .press: return 1
        case .repeat: return 2
        }
    }

    static func mods(for value: TerminalKeyMods) -> Int {
        var bits = 0
        if value.shift { bits |= 1 << 0 }
        if value.ctrl { bits |= 1 << 1 }
        if value.alt { bits |= 1 << 2 }
        if value.meta { bits |= 1 << 3 }
        return bits
    }

    static func litterKey(for value: TerminalKeyCode) -> LitterGhosttyKey {
        switch value {
        case .enter: return .enter
        case .tab: return .tab
        case .backspace: return .backspace
        case .escape: return .escape
        case .space: return .space
        case .arrowUp: return .arrowUp
        case .arrowDown: return .arrowDown
        case .arrowLeft: return .arrowLeft
        case .arrowRight: return .arrowRight
        case .pageUp: return .pageUp
        case .pageDown: return .pageDown
        case .home: return .home
        case .end: return .end
        case .delete: return .delete
        case .insert: return .insert
        default: return .unidentified
        }
    }
}

File: apps/android/app/src/main/java/com/litter/android/ui/terminal/GhosttyRendererBackendBridge.kt
package com.litter.android.ui.terminal

import android.os.Handler
import android.os.Looper
import com.litter.android.core.bridge.GhosttyRendererBridge
import uniffi.codex_mobile_client.TerminalCellMetrics
import uniffi.codex_mobile_client.TerminalCellRange
import uniffi.codex_mobile_client.TerminalKeyAction
import uniffi.codex_mobile_client.TerminalKeyCode
import uniffi.codex_mobile_client.TerminalKeyEvent
import uniffi.codex_mobile_client.TerminalKeyMods
import uniffi.codex_mobile_client.TerminalRendererBackend

/**
 * Kotlin implementation of the Rust-defined `TerminalRendererBackend` callback
 * interface. The Rust [`uniffi.codex_mobile_client.TerminalRenderer`] tick task
 * invokes these methods from a tokio worker; we hop to the main thread before
 * touching the (non-thread-safe) Ghostty surface APIs.
 */
internal class GhosttyRendererBackendBridge(
    private val surface: GhosttyRendererBridge.GhosttyRendererSurface,
    private val onRequestRedraw: () -> Unit,
) : TerminalRendererBackend {

    private val mainHandler = Handler(Looper.getMainLooper())

    override fun setFocus(focused: Boolean) {
        runOnMain { surface.setFocus(focused) }
    }

    override fun setOcclusion(occluded: Boolean) {
        runOnMain { surface.setOcclusion(occluded) }
    }

    override fun requestRedraw() {
        runOnMain { onRequestRedraw() }
    }

    override fun applyConfigFile(path: String) {
        runOnMain { surface.applyConfig(path) }
    }

    override fun dispatchKey(event: TerminalKeyEvent) {
        val action = when (event.action) {
            TerminalKeyAction.RELEASE -> 0
            TerminalKeyAction.PRESS -> 1
            TerminalKeyAction.REPEAT -> 2
        }
        val key = bridgeKey(event.code)
        val mods = packMods(event.mods)
        val text = event.text.ifEmpty { null }
        runOnMain { surface.sendKey(action, key, mods, text, composing = false) }
    }

    override fun dispatchText(text: String, composing: Boolean) {
        runOnMain {
            if (composing) {
                surface.sendPreedit(text.ifEmpty { null })
            } else if (text.isNotEmpty()) {
                surface.sendText(text)
            }
        }
    }

    override fun dispatchPaste(bytes: ByteArray) {
        // Bracketed paste bytes are valid PTY input — write them directly so
        // Ghostty's input parser sees the wrapper unmodified.
        runOnMain { surface.write(bytes) }
    }

    // Selection support — Rust drives word/line range math + read-back, but
    // the painted overlay + real ghostty_surface_read_selection wiring is
    // a follow-up PR. For now return defaults so the trait satisfies and
    // selection_set is observable when the overlay lands.
    override fun readSelection(): String? = null

    override fun readText(
        startRow: UInt,
        startCol: UInt,
        endRow: UInt,
        endCol: UInt,
    ): String? = null

    override fun cellMetrics(): TerminalCellMetrics = TerminalCellMetrics(
        cellWidthPx = 0f,
        cellHeightPx = 0f,
        cols = 0u,
        rows = 0u,
        viewportTop = 0u,
    )

    override fun setSelectionOverlay(range: TerminalCellRange?) {
        // Overlay painting arrives in the follow-up PR.
    }

    private fun packMods(mods: TerminalKeyMods): Int {
        var bits = 0
        if (mods.shift) bits = bits or (1 shl 0)
        if (mods.ctrl) bits = bits or (1 shl 1)
        if (mods.alt) bits = bits or (1 shl 2)
        if (mods.meta) bits = bits or (1 shl 3)
        return bits
    }

    // Mirrors `LitterBridgeKey` in ghostty_jni.cpp; the JNI bridge does the
    // final translation to ghostty_input_key_e.
    private fun bridgeKey(code: TerminalKeyCode): Int = when (code) {
        is TerminalKeyCode.Enter -> 1
        is TerminalKeyCode.Tab -> 2
        is TerminalKeyCode.Backspace -> 3
        is TerminalKeyCode.Escape -> 4
        is TerminalKeyCode.Space -> 5
        is TerminalKeyCode.ArrowUp -> 6
        is TerminalKeyCode.ArrowDown -> 7
        is TerminalKeyCode.ArrowLeft -> 8
        is TerminalKeyCode.ArrowRight -> 9
        is TerminalKeyCode.PageUp -> 10
        is TerminalKeyCode.PageDown -> 11
        is TerminalKeyCode.Home -> 12
        is TerminalKeyCode.End -> 13
        is TerminalKeyCode.Delete -> 14
        is TerminalKeyCode.Insert -> 15
        else -> 0
    }

    private fun runOnMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            mainHandler.post(block)
        }
    }
}
